File: xy/web/ts/rank.ts
// rank.ts — fractional indexing (LexoRank-style) so a drag updates only the
// moved item's rank. A faithful port of the public-domain `fractional-indexing`
// algorithm (rocicorp/fractional-indexing, MIT). Keys are base-62 strings that
// sort lexicographically; `keyBetween(a, b)` returns a key strictly between a
// and b (null = open end).

const DIGITS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
const ZERO = DIGITS[0];

function midpoint(a: string, b: string | null): string {
  if (b !== null && a >= b) throw new Error(`${a} >= ${b}`);
  if (a.slice(-1) === ZERO || (b && b.slice(-1) === ZERO)) throw new Error("trailing zero");
  if (b) {
    let n = 0;
    while ((a[n] || ZERO) === b[n]) n++;
    if (n > 0) return b.slice(0, n) + midpoint(a.slice(n), b.slice(n));
  }
  const digitA = a ? DIGITS.indexOf(a[0]) : 0;
  const digitB = b !== null ? DIGITS.indexOf(b[0]) : DIGITS.length;
  if (digitB - digitA > 1) {
    const midDigit = Math.round(0.5 * (digitA + digitB));
    return DIGITS[midDigit];
  }
  if (b && b.length > 1) return b.slice(0, 1);
  return DIGITS[digitA] + midpoint(a.slice(1), null);
}

function validateInteger(int: string): void {
  if (int.length !== getIntegerLength(int[0])) throw new Error(`invalid integer part: ${int}`);
}
function getIntegerLength(head: string): number {
  if (head >= "a" && head <= "z") return head.charCodeAt(0) - "a".charCodeAt(0) + 2;
  if (head >= "A" && head <= "Z") return "Z".charCodeAt(0) - head.charCodeAt(0) + 2;
  throw new Error("invalid order key head: " + head);
}
function getIntegerPart(key: string): string {
  const integerPartLength = getIntegerLength(key[0]);
  if (integerPartLength > key.length) throw new Error("invalid order key: " + key);
  return key.slice(0, integerPartLength);
}
function validateOrderKey(key: string): void {
  if (key === "A" + ZERO.repeat(26)) throw new Error("invalid order key: " + key);
  const i = getIntegerPart(key);
  const f = key.slice(i.length);
  if (f.slice(-1) === ZERO) throw new Error("invalid order key: " + key);
}

function incrementInteger(x: string): string | null {
  validateInteger(x);
  const [head, ...digs] = x.split("");
  let carry = true;
  for (let i = digs.length - 1; carry && i >= 0; i--) {
    const d = DIGITS.indexOf(digs[i]) + 1;
    if (d === DIGITS.length) digs[i] = ZERO;
    else { digs[i] = DIGITS[d]; carry = false; }
  }
  if (carry) {
    if (head === "Z") return "a" + ZERO;
    if (head === "z") return null;
    const h = String.fromCharCode(head.charCodeAt(0) + 1);
    if (h > "a") digs.push(ZERO);
    else digs.shift();
    return h + digs.join("");
  }
  return head + digs.join("");
}
function decrementInteger(x: string): string | null {
  validateInteger(x);
  const [head, ...digs] = x.split("");
  let borrow = true;
  for (let i = digs.length - 1; borrow && i >= 0; i--) {
    const d = DIGITS.indexOf(digs[i]) - 1;
    if (d === -1) digs[i] = DIGITS.slice(-1);
    else { digs[i] = DIGITS[d]; borrow = false; }
  }
  if (borrow) {
    if (head === "a") return "Z" + DIGITS.slice(-1);
    if (head === "A") return null;
    const h = String.fromCharCode(head.charCodeAt(0) - 1);
    if (h < "Z") digs.push(DIGITS.slice(-1));
    else digs.shift();
    return h + digs.join("");
  }
  return head + digs.join("");
}

// keyBetween returns a key strictly between a and b (null open ends).
function keyBetween(a: string | null, b: string | null): string {
  if (a !== null) validateOrderKey(a);
  if (b !== null) validateOrderKey(b);
  if (a !== null && b !== null && a >= b) throw new Error(`${a} >= ${b}`);
  if (a === null) {
    if (b === null) return "a" + ZERO;
    const ib = getIntegerPart(b);
    const fb = b.slice(ib.length);
    if (ib === "A" + ZERO.repeat(26)) return ib + midpoint("", fb);
    if (ib < b) return ib;
    const res = decrementInteger(ib);
    if (res === null) throw new Error("cannot decrement any more");
    return res;
  }
  if (b === null) {
    const ia = getIntegerPart(a);
    const fa = a.slice(ia.length);
    const i = incrementInteger(ia);
    return i === null ? ia + midpoint(fa, null) : i;
  }
  const ia = getIntegerPart(a);
  const fa = a.slice(ia.length);
  const ib = getIntegerPart(b);
  const fb = b.slice(ib.length);
  if (ia === ib) return ia + midpoint(fa, fb);
  const i = incrementInteger(ia);
  if (i === null) throw new Error("cannot increment any more");
  if (i < b) return i;
  return ia + midpoint(fa, null);
}

// nKeysBetween returns n evenly distributed keys strictly between a and b.
function nKeysBetween(a: string | null, b: string | null, n: number): string[] {
  if (n === 0) return [];
  if (n === 1) return [keyBetween(a, b)];
  if (b === null) {
    let c = keyBetween(a, b);
    const out = [c];
    for (let i = 0; i < n - 1; i++) { c = keyBetween(c, b); out.push(c); }
    return out;
  }
  if (a === null) {
    let c = keyBetween(a, b);
    const out = [c];
    for (let i = 0; i < n - 1; i++) { c = keyBetween(a, c); out.push(c); }
    out.reverse();
    return out;
  }
  const mid = Math.floor(n / 2);
  const c = keyBetween(a, b);
  return [...nKeysBetween(a, c, mid), c, ...nKeysBetween(c, b, n - mid - 1)];
}

export const xyRank = { keyBetween, nKeysBetween };

File: xy/internal/rank/rank.go
// Package rank is fractional indexing: base-62 keys that sort
// lexicographically, so moving one card rewrites one row. A faithful port of
// the public-domain `fractional-indexing` algorithm (rocicorp/fractional-indexing,
// MIT), the same one web/ts/rank.ts runs in the browser — the two must agree,
// or a card lands in a different place depending on who moved it.
//
// An open end is the empty string here, where the TS uses null.
package rank

import (
	"errors"
	"strings"
)

const digits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
const zero = '0'

func indexOf(c byte) int { return strings.IndexByte(digits, c) }

// midpoint is the shortest string strictly between the fractional parts a and b
// (b empty = open right end).
func midpoint(a, b string) (string, error) {
	if b != "" && a >= b {
		return "", errors.New("rank: " + a + " >= " + b)
	}
	if strings.HasSuffix(a, string(zero)) || strings.HasSuffix(b, string(zero)) {
		return "", errors.New("rank: trailing zero")
	}
	if b != "" {
		n := 0
		for n < len(b) {
			ca := byte(zero)
			if n < len(a) {
				ca = a[n]
			}
			if ca != b[n] {
				break
			}
			n++
		}
		if n > 0 {
			restA := ""
			if n < len(a) {
				restA = a[n:]
			}
			rest, err := midpoint(restA, b[n:])
			if err != nil {
				return "", err
			}
			return b[:n] + rest, nil
		}
	}
	digitA := 0
	if a != "" {
		digitA = indexOf(a[0])
	}
	digitB := len(digits)
	if b != "" {
		digitB = indexOf(b[0])
	}
	if digitB-digitA > 1 {
		return string(digits[(digitA+digitB+1)/2]), nil // == Math.round(0.5*(a+b))
	}
	if len(b) > 1 {
		return b[:1], nil
	}
	rest := ""
	if len(a) > 0 {
		rest = a[1:]
	}
	tail, err := midpoint(rest, "")
	if err != nil {
		return "", err
	}
	return string(digits[digitA]) + tail, nil
}

func integerLength(head byte) (int, error) {
	switch {
	case head >= 'a' && head <= 'z':
		return int(head-'a') + 2, nil
	case head >= 'A' && head <= 'Z':
		return int('Z'-head) + 2, nil
	}
	return 0, errors.New("rank: invalid order key head: " + string(head))
}

func integerPart(key string) (string, error) {
	if key == "" {
		return "", errors.New("rank: empty order key")
	}
	n, err := integerLength(key[0])
	if err != nil {
		return "", err
	}
	if n > len(key) {
		return "", errors.New("rank: invalid order key: " + key)
	}
	return key[:n], nil
}

func validate(key string) error {
	if key == "A"+strings.Repeat(string(zero), 26) {
		return errors.New("rank: invalid order key: " + key)
	}
	i, err := integerPart(key)
	if err != nil {
		return err
	}
	if strings.HasSuffix(key[len(i):], string(zero)) {
		return errors.New("rank: invalid order key: " + key)
	}
	return nil
}

// incrementInteger returns the next integer part; ok=false when there is none
// ('z' fully carried), which the caller answers by extending the fraction.
func incrementInteger(x string) (string, bool, error) {
	if err := validateInteger(x); err != nil {
		return "", false, err
	}
	head, digs := x[0], []byte(x[1:])
	carry := true
	for i := len(digs) - 1; carry && i >= 0; i-- {
		if d := indexOf(digs[i]) + 1; d == len(digits) {
			digs[i] = zero
		} else {
			digs[i] = digits[d]
			carry = false
		}
	}
	if carry {
		switch head {
		case 'Z':
			return "a" + string(zero), true, nil
		case 'z':
			return "", false, nil
		}
		h := head + 1
		if h > 'a' {
			digs = append(digs, zero)
		} else if len(digs) > 0 {
			digs = digs[1:]
		}
		return string(h) + string(digs), true, nil
	}
	return string(head) + string(digs), true, nil
}

// decrementInteger is incrementInteger's mirror; ok=false at 'A', the low end.
func decrementInteger(x string) (string, bool, error) {
	if err := validateInteger(x); err != nil {
		return "", false, err
	}
	head, digs := x[0], []byte(x[1:])
	last := digits[len(digits)-1]
	borrow := true
	for i := len(digs) - 1; borrow && i >= 0; i-- {
		if d := indexOf(digs[i]) - 1; d == -1 {
			digs[i] = last
		} else {
			digs[i] = digits[d]
			borrow = false
		}
	}
	if borrow {
		switch head {
		case 'a':
			return "Z" + string(last), true, nil
		case 'A':
			return "", false, nil
		}
		h := head - 1
		if h < 'Z' {
			digs = append(digs, last)
		} else if len(digs) > 0 {
			digs = digs[1:]
		}
		return string(h) + string(digs), true, nil
	}
	return string(head) + string(digs), true, nil
}

func validateInteger(x string) error {
	if x == "" {
		return errors.New("rank: empty integer part")
	}
	n, err := integerLength(x[0])
	if err != nil {
		return err
	}
	if len(x) != n {
		return errors.New("rank: invalid integer part: " + x)
	}
	return nil
}

// Between returns a key strictly between a and b. Either may be "" for an open
// end, so Between("", "") is the first key of an empty list.
func Between(a, b string) (string, error) {
	if a != "" {
		if err := validate(a); err != nil {
			return "", err
		}
	}
	if b != "" {
		if err := validate(b); err != nil {
			return "", err
		}
	}
	if a != "" && b != "" && a >= b {
		return "", errors.New("rank: " + a + " >= " + b)
	}
	switch {
	case a == "" && b == "":
		return "a" + string(zero), nil
	case a == "":
		ib, err := integerPart(b)
		if err != nil {
			return "", err
		}
		fb := b[len(ib):]
		if ib == "A"+strings.Repeat(string(zero), 26) {
			mid, err := midpoint("", fb)
			if err != nil {
				return "", err
			}
			return ib + mid, nil
		}
		if ib < b {
			return ib, nil
		}
		res, ok, err := decrementInteger(ib)
		if err != nil {
			return "", err
		}
		if !ok {
			return "", errors.New("rank: cannot decrement any more")
		}
		return res, nil
	case b == "":
		ia, err := integerPart(a)
		if err != nil {
			return "", err
		}
		fa := a[len(ia):]
		i, ok, err := incrementInteger(ia)
		if err != nil {
			return "", err
		}
		if ok {
			return i, nil
		}
		mid, err := midpoint(fa, "")
		if err != nil {
			return "", err
		}
		return ia + mid, nil
	}
	ia, err := integerPart(a)
	if err != nil {
		return "", err
	}
	ib, err := integerPart(b)
	if err != nil {
		return "", err
	}
	fa, fb := a[len(ia):], b[len(ib):]
	if ia == ib {
		mid, err := midpoint(fa, fb)
		if err != nil {
			return "", err
		}
		return ia + mid, nil
	}
	i, ok, err := incrementInteger(ia)
	if err != nil {
		return "", err
	}
	if !ok {
		return "", errors.New("rank: cannot increment any more")
	}
	if i < b {
		return i, nil
	}
	mid, err := midpoint(fa, "")
	if err != nil {
		return "", err
	}
	return ia + mid, nil
}

// After appends: a key strictly greater than prev, or the first key when prev
// is "". What the Trello-compatible upload ranks a new card by.
func After(prev string) (string, error) { return Between(prev, "") }
